## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [143]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [144]:
# TODO: load `data/processed/shopping_cleaned.csv` as a pandas dataframe

df = pd.read_csv('../data/processed/shopping_cleaned.csv')
df

Age  Gender Item Purchased  Purchase Amount (USD)      Location Size  \
0     31.0    Male       Leggings              24.231704        Nevada    M   
1     38.0    Male     Sunglasses              36.545487        Oregon    S   
2     26.0    Male       Leggings              23.595139        Nevada   XL   
3     43.0    Male          Dress              34.079015    California    M   
4     29.0    Male         Shorts              23.796439     Minnesota    M   
...    ...     ...            ...                    ...           ...  ...   
3153  39.0    Male  Running Shoes              55.636680        Kansas    L   
3154  43.0  Female     Sunglasses              61.610602      Colorado    S   
3155  37.0    Male        Pajamas              44.600556        Alaska    S   
3156  39.0    Male         Hoodie              45.343778      Illinois    S   
3157  21.0  Female         Gloves              49.439181  North Dakota    M   

             Color  Season Review Rating   Shipping Type Promo Code Used  \
0      Terra cotta  Winter       Present        Standard              No   
1        Aubergine  Summer       Missing        Standard              No   
2            Brown  Winter       Missing        Standard              No   
3      Terra cotta    Fall       Present        Standard              No   
4         Lavender  Summer       Present         Express              No   
...            ...     ...           ...             ...             ...   
3153         White  Summer       Missing  2-Day Shipping             Yes   
3154  Burnt orange    Fall       Missing  2-Day Shipping             Yes   
3155     Aubergine  Winter       Present  2-Day Shipping             Yes   
3156     Brick red    Fall       Missing  2-Day Shipping             Yes   
3157     Aubergine  Winter       Present  2-Day Shipping             Yes   

      Previous Purchases Payment Method  
0                      0    Credit Card  
1                      0    Credit Card  
2                      3    Credit Card  
3                      0    Credit Card  
4                      0    Credit Card  
...                  ...            ...  
3153                  11          Venmo  
3154                   7          Venmo  
3155                  20          Venmo  
3156                   3          Venmo  
3157                   2          Venmo  

[3158 rows x 13 columns]

In [145]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

color_by_season_table = pd.pivot_table(df, values='Location', index='Color', columns='Season', aggfunc=len)

In [146]:
# TODO: Display this pivot table

color_by_season_table

Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine               83       2      13     111
Baby blue               16      69      84      21
Black                   84       5       9     124
Brick red               88       7      12     110
Brown                   97       5       3      89
Burnt orange            93       4       5     112
Fuchsia                  3      45      78      18
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Mauve                   86       5       7      95
Muted mustard yellow    84       8      13      87
Pale peach               9      52      81      15
Periwinkle               9      54      82       8
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Turquoise                7      49      74      16
White                   10      63      81      11

In [147]:
# TODO: Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

color_by_season_table.sort_values(['Fall'], ascending=False)

Season                Fall  Spring  Summer  Winter
Color                                             
Brown                   97       5       3      89
Burnt orange            93       4       5     112
Terra cotta             91       4      15     101
Ruby Red                90       3       8     107
Brick red               88       7      12     110
Mauve                   86       5       7      95
Black                   84       5       9     124
Muted mustard yellow    84       8      13      87
Aubergine               83       2      13     111
Baby blue               16      69      84      21
Lavender                13      55     102       6
White                   10      63      81      11
Periwinkle               9      54      82       8
Pale peach               9      52      81      15
Lemon yellow             9      47      98      13
Turquoise                7      49      74      16
Fuchsia                  3      45      78      18

In [148]:
# TODO: Select the "Winter" column from this pivot table and display the sort order in descending order

color_by_season_table.sort_values('Winter', ascending=False)

Season                Fall  Spring  Summer  Winter
Color                                             
Black                   84       5       9     124
Burnt orange            93       4       5     112
Aubergine               83       2      13     111
Brick red               88       7      12     110
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Mauve                   86       5       7      95
Brown                   97       5       3      89
Muted mustard yellow    84       8      13      87
Baby blue               16      69      84      21
Fuchsia                  3      45      78      18
Turquoise                7      49      74      16
Pale peach               9      52      81      15
Lemon yellow             9      47      98      13
White                   10      63      81      11
Periwinkle               9      54      82       8
Lavender                13      55     102       6

In [149]:
# TODO: Select the "Spring" column from this pivot table and display the sort order in descending order

color_by_season_table.sort_values('Spring', ascending=False)

Season                Fall  Spring  Summer  Winter
Color                                             
Baby blue               16      69      84      21
White                   10      63      81      11
Lavender                13      55     102       6
Periwinkle               9      54      82       8
Pale peach               9      52      81      15
Turquoise                7      49      74      16
Lemon yellow             9      47      98      13
Fuchsia                  3      45      78      18
Muted mustard yellow    84       8      13      87
Brick red               88       7      12     110
Mauve                   86       5       7      95
Brown                   97       5       3      89
Black                   84       5       9     124
Burnt orange            93       4       5     112
Terra cotta             91       4      15     101
Ruby Red                90       3       8     107
Aubergine               83       2      13     111

In [150]:
# TODO: Select the "Summer" column from this pivot table and display the sort order in descending order

color_by_season_table.sort_values('Summer', ascending=False)

Season                Fall  Spring  Summer  Winter
Color                                             
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Baby blue               16      69      84      21
Periwinkle               9      54      82       8
Pale peach               9      52      81      15
White                   10      63      81      11
Fuchsia                  3      45      78      18
Turquoise                7      49      74      16
Terra cotta             91       4      15     101
Muted mustard yellow    84       8      13      87
Aubergine               83       2      13     111
Brick red               88       7      12     110
Black                   84       5       9     124
Ruby Red                90       3       8     107
Mauve                   86       5       7      95
Burnt orange            93       4       5     112
Brown                   97       5       3      89

In [151]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

color_by_season_table.idxmax(axis=0)

Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

In [152]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

item_purchased_by_season_table = pd.pivot_table(df, values='Location', index='Item Purchased', columns='Season', aggfunc=len)

item_purchased_by_season_table

Season          Fall  Spring  Summer  Winter
Item Purchased                              
Backpack        72.0     NaN     NaN   113.0
Belt            25.0     NaN     NaN     NaN
Boots           51.0     NaN     NaN    83.0
Dress           36.0    74.0   118.0     NaN
Gloves          68.0     NaN     NaN   112.0
Handbag         31.0    78.0   111.0     NaN
Hoodie          58.0     NaN     NaN   105.0
Jacket          47.0     NaN     NaN    87.0
Jeans           46.0     NaN     NaN     NaN
Leggings        58.0     NaN     NaN   118.0
Pajamas         65.0     NaN     NaN   111.0
Running Shoes   25.0    89.0   134.0     NaN
Scarf           56.0     NaN     NaN   101.0
Shorts          31.0    76.0   152.0     NaN
Socks           71.0     NaN     NaN   101.0
Sunglasses      43.0    80.0   135.0     NaN
Sweater         72.0     NaN     NaN   113.0
T-shirt         17.0    80.0   115.0     NaN

In [153]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

item_purchased_by_season_table.idxmax()


Season
Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object

In [154]:
# TODO: Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two dataframes into seperate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

promo_code_users = pd.DataFrame(df[df['Promo Code Used'] == 'Yes'])
non_promo_code_users = pd.DataFrame(df[df['Promo Code Used'] == 'No'])

In [155]:
# TODO: Print the shape of the promo code users dataframe. This should be (1373, 13)

print(promo_code_users.shape)

(1373, 13)


In [156]:
# TODO: Print the average purchase size of promo code users dataframe. This should be 50.06747998236351

print(np.mean(promo_code_users['Purchase Amount (USD)']))

50.06747998236351


In [157]:
# TODO: Print the shape of the non promo code users dataframe. This should be (1785, 13)

print(non_promo_code_users.shape)

(1785, 13)


In [158]:
# TODO: the average purchase size of non promo code users dataframe. This should be 30.155750765104113

print(np.mean(non_promo_code_users['Purchase Amount (USD)']))

30.155750765104113


Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

The null hypothesis states that there is no difference between the dollar amount spent between promo code users and non-promo-code users.

Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

The alternative hypothesis states that there is a difference between the dollar amount spent between promo code users and non-promo-code users.

In [159]:
# TODO: Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

stats.ttest_ind(promo_code_users['Purchase Amount (USD)'], non_promo_code_users['Purchase Amount (USD)'])

TtestResult(statistic=93.2464714350831, pvalue=0.0, df=3156.0)

In [160]:
# TODO: Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

g_review = df.groupby('Review Rating')

In [161]:
# TODO: Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html

g_review.groups.keys() 

dict_keys(['Missing', 'Present'])

In [162]:
# TODO: calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

g_review[['Purchase Amount (USD)', 'Age', 'Previous Purchases']].mean()

Purchase Amount (USD)        Age  Previous Purchases
Review Rating                                                      
Missing                    38.923623  34.080198            5.344059
Present                    38.615947  34.371705            6.778559

In [163]:
# TODO: Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

df['Prev Purchase Group'] = pd.cut(df['Previous Purchases'], bins=2, labels=['Low','High'])


In [164]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

low_group = df[df['Prev Purchase Group'] == 'Low'].value_counts('Review Rating')
low_group

Review Rating
Missing    1686
Present     840
Name: count, dtype: int64

In [165]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

high_group = df[df['Prev Purchase Group'] == 'High'].value_counts('Review Rating')
high_group


Review Rating
Missing    334
Present    298
Name: count, dtype: int64

In [166]:
#Proportions of low_group
low_group = df[df['Prev Purchase Group'] == 'Low'].value_counts('Review Rating', normalize=True)
low_group

Review Rating
Missing    0.667458
Present    0.332542
Name: proportion, dtype: float64

In [167]:
#Proportions of high_group
high_group = df[df['Prev Purchase Group'] == 'High'].value_counts('Review Rating', normalize=True)
high_group

Review Rating
Missing    0.528481
Present    0.471519
Name: proportion, dtype: float64

## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

Top 3 colors for Fall:
Baby blue, White, Lavender
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Season</th>
      <th>Fall</th>
      <th>Spring</th>
      <th>Summer</th>
      <th>Winter</th>
    </tr>
    <tr>
      <th>Color</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Brown</th>
      <td>97</td>
      <td>5</td>
      <td>3</td>
      <td>89</td>
    </tr>
    <tr>
      <th>Burnt orange</th>
      <td>93</td>
      <td>4</td>
      <td>5</td>
      <td>112</td>
    </tr>
    <tr>
      <th>Terra cotta</th>
      <td>91</td>
      <td>4</td>
      <td>15</td>
      <td>101</td>
    </tr>
  </tbody>
</table>
</div>

Top 3 colors for Winter:
Black, Burnt orange, Aubergine
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Season</th>
      <th>Fall</th>
      <th>Spring</th>
      <th>Summer</th>
      <th>Winter</th>
    </tr>
    <tr>
      <th>Color</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Black</th>
      <td>84</td>
      <td>5</td>
      <td>9</td>
      <td>124</td>
    </tr>
    <tr>
      <th>Burnt orange</th>
      <td>93</td>
      <td>4</td>
      <td>5</td>
      <td>112</td>
    </tr>
    <tr>
      <th>Aubergine</th>
      <td>83</td>
      <td>2</td>
      <td>13</td>
      <td>111</td>
    </tr>
  </tbody>
</table>
</div>

### Q2

What are the top three colors for Spring & Summer?

Top 3 colors for Spring:
Baby blue, White, Lavender
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Season</th>
      <th>Fall</th>
      <th>Spring</th>
      <th>Summer</th>
      <th>Winter</th>
    </tr>
    <tr>
      <th>Color</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Baby blue</th>
      <td>16</td>
      <td>69</td>
      <td>84</td>
      <td>21</td>
    </tr>
    <tr>
      <th>White</th>
      <td>10</td>
      <td>63</td>
      <td>81</td>
      <td>11</td>
    </tr>
    <tr>
      <th>Lavender</th>
      <td>13</td>
      <td>55</td>
      <td>102</td>
      <td>6</td>
    </tr>
  </tbody>
</table>
</div>

Top 3 colors for Summer:
Lavender, Lemon yellow, Baby blue
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>Season</th>
      <th>Fall</th>
      <th>Spring</th>
      <th>Summer</th>
      <th>Winter</th>
    </tr>
    <tr>
      <th>Color</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Lavender</th>
      <td>13</td>
      <td>55</td>
      <td>102</td>
      <td>6</td>
    </tr>
    <tr>
      <th>Lemon yellow</th>
      <td>9</td>
      <td>47</td>
      <td>98</td>
      <td>13</td>
    </tr>
    <tr>
      <th>Baby blue</th>
      <td>16</td>
      <td>69</td>
      <td>84</td>
      <td>21</td>
    </tr>
  </tbody>
</table>
</div>

color_by_season_table.sort_values('...', ascending=False).head(3)

to create pivot tables with top 3

### Q3

What is the most popular clothing item by season?

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Item Purchased</th>
    </tr>
    <tr>
      <th>Season</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Fall</th>
      <td>Backpack</td>
    </tr>
    <tr>
      <th>Spring</th>
      <td>Running Shoes</td>
    </tr>
    <tr>
      <th>Summer</th>
      <td>Shorts</td>
    </tr>
    <tr>
      <th>Winter</th>
      <td>Leggings</td>
    </tr>
  </tbody>
</table>
</div>

item_purchased_by_season_subtable = pd.DataFrame(item_purchased_by_season_table.idxmax())

item_purchased_by_season_subtable.rename(columns={0: 'Item Purchased'})

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

We received a p-value of 0.0. If our cutoff is 0.05, we can therefore reject the null hypothesis.

0.0 < 0.05

*TtestResult(statistic=93.2464714350831, pvalue=0.0, df=3156.0)*

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

Although there is a higher total number of reviews that are present in the Low Group at 640, the proportion of reviews that are present in the High Group is higher at 47%. My intuition leads me to believe that people tend to leave reviews when they feel polarized about an item. That is they feel strongly positive or strongly negative about an item.


*Low Group proportion output:*

*Missing    1686 (0.667458)*

*Present    640 (0.332542)**

*High Group proportion output:*

*Missing    334 (0.528481)*

*Present    298 (0.471519)**

